python/amici/pysb_import.py

```diff
@@ -240,11 +240,111 @@ def ode_model_from_pysb_importer(
         for noise_distr in noise_distributions.values()
     )
 
+    _process_stoichiometric_matrix(model, ode, constant_parameters)
+
     ode.generate_basic_variables()
 
     return ode
 
 
+@log_execution_time('processing PySB stoich. matrix', logger)
+def _process_stoichiometric_matrix(pysb_model: pysb.Model,
+                                   ode_model: ODEModel,
+                                   constant_parameters: List[str]) -> None:
+
+    """
+    Exploits the PySB stoichiometric matrix to generate xdot derivatives
+
+    :param pysb_model:
+        pysb model instance
+
+    :param ode_model:
+        ODEModel instance
+
+    :param constant_parameters:
+        list of constant parameters
+    """
+
+    x = ode_model.sym('x')
+    w = list(ode_model.sym('w'))
+    p = list(ode_model.sym('p'))
+    x_rdata = list(ode_model.sym('x_rdata'))
+
+    n_x = len(x)
+    n_w = len(w)
+    n_p = len(p)
+    n_r = len(pysb_model.reactions)
+
+    solver_index = ode_model.get_solver_indices()
+    dflux_dx_dict = {}
+    dflux_dw_dict = {}
+    dflux_dp_dict = {}
+
+    w_idx = dict()
+    p_idx = dict()
+    wx_idx = dict()
+
+    def get_cached_index(symbol, sarray, index_cache):
+        if symbol not in index_cache:
+            idx = sarray.index(symbol)
+            index_cache[symbol] = idx
+        else:
+            idx = index_cache[symbol]
+        return idx
```

Comment:
```suggestion
        idx = index_cache.get(symbol, None)
        if idx is not None:
            return idx
        idx = sarray.index(symbol)
        index_cache[symbol] = idx
        return idx
```

Avoids looking up `symbol` in `index_cache` twice